File: lib/flow_regex/parser.rb
class FlowRegex
  # 正規表現パーサー
  # 基本的な再帰下降パーサーで連接、選択、クリーネ閉包をサポート
  class Parser
    def initialize(pattern)
      @pattern = pattern
      @pos = 0
    end
    
    def parse
      result = parse_expression
      raise ParseError, "Unexpected character at position #{@pos}: '#{current_char}'" unless at_end?
      result
    end
    
    private
    
    # Expression := Term ('|' Term)*
    def parse_expression
      left = parse_term
      
      while current_char == '|'
        consume('|')
        right = parse_term
        left = Alternation.new(left, right)
      end
      
      left
    end
    
    # Term := Factor*
    def parse_term
      factors = []
      
      while !at_end? && current_char != '|' && current_char != ')'
        factors << parse_factor
      end
      
      case factors.length
      when 0
        raise ParseError, "Empty term at position #{@pos}"
      when 1
        factors.first
      else
        # 複数の要素を左結合で連接
        factors.reduce { |acc, factor| Concat.new(acc, factor) }
      end
    end
    
    # Factor := Atom ('*' | '+' | '?' | '{' NUMBER (',' NUMBER?)? '}')?
    def parse_factor
      atom = parse_atom
      
      case current_char
      when '*'
        consume('*')
        KleeneStar.new(atom)
      when '+'
        consume('+')
        Plus.new(atom)
      when '?'
        consume('?')
        Question.new(atom)
      when '{'
        parse_quantifier(atom)
      else
        atom
      end
    end
    
    # 量指定子 {n} または {n,m} をパース
    def parse_quantifier(atom)
      consume('{')
      
      # 最初の数値を読む
      min_count = parse_number
      
      if current_char == ','
        consume(',')
        if current_char == '}'
          # {n,} の形式（n回以上）
          consume('}')
          RangeCount.new(atom, min_count, -1)  # -1 は無限大
        else
          # {n,m} の形式
          max_count = parse_number
          consume('}')
          if min_count > max_count
            raise ParseError, "Invalid quantifier range: {#{min_count},#{max_count}} at position #{@pos}"
          end
          RangeCount.new(atom, min_count, max_count)
        end
      else
        # {n} の形式（n回ちょうど）
        consume('}')
        ExactCount.new(atom, min_count)
      end
    end
    
    # 数値をパース
    def parse_number
      start_pos = @pos
      
      unless current_char && current_char.match(/\d/)
        raise ParseError, "Expected number at position #{@pos}"
      end
      
      while current_char && current_char.match(/\d/)
        advance
      end
      
      @pattern[start_pos...@pos].to_i
    end
    
    # Atom := CHAR | '(' Expression ')' | '[' CharClass ']' | '\' EscapeChar | '.'
    def parse_atom
      case current_char
      when '('
        consume('(')
        expr = parse_expression
        consume(')')
        expr
      when '['
        parse_character_class
      when '\\'
        parse_escape_sequence
      when '.'
        advance
        AnyChar.new
      when nil
        raise ParseError, "Unexpected end of pattern"
      when '|', ')', '*', '+', '?', '{', ']'
        raise ParseError, "Unexpected character '#{current_char}' at position #{@pos}"
      else
        char = current_char
        advance
        Literal.new(char)
      end
    end
    
    # 文字クラス [a-z] をパース
    def parse_character_class
      consume('[')
      
      start_pos = @pos
      class_content = ""
      
      # ] までの内容を読む
      while current_char && current_char != ']'
        class_content += current_char
        advance
      end
      
      if current_char != ']'
        raise ParseError, "Unclosed character class at position #{start_pos}"
      end
      
      consume(']')
      
      if class_content.empty?
        raise ParseError, "Empty character class at position #{start_pos}"
      end
      
      CharacterClass.new(class_content)
    end
    
    # エスケープシーケンス \d, \n などをパース
    def parse_escape_sequence
      consume('\\')
      
      if current_char.nil?
        raise ParseError, "Incomplete escape sequence at end of pattern"
      end
      
      escape_char = current_char
      advance
      
      case escape_char
      when 'd', 'D', 's', 'S', 'w', 'W'
        # 文字クラス略記
        CharacterClass.new(escape_char)
      else
        # 通常のエスケープシーケンス
        EscapeSequence.new(escape_char)
      end
    end
    
    def current_char
      return nil if @pos >= @pattern.length
      @pattern[@pos]
    end
    
    def advance
      @pos += 1
    end
    
    def consume(expected)
      if current_char == expected
        advance
      else
        raise ParseError, "Expected '#{expected}' at position #{@pos}, got '#{current_char}'"
      end
    end
    
    def at_end?
      @pos >= @pattern.length
    end
  end
end

File: c_implementation/src/parser.c
#include "flowregex.h"
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <ctype.h>

typedef struct {
    const char *pattern;
    size_t pos;
    size_t length;
    flowregex_error_t *error;
} parser_state_t;

// Forward declarations
static regex_element_t *parse_expression(parser_state_t *state);
static regex_element_t *parse_term(parser_state_t *state);
static regex_element_t *parse_factor(parser_state_t *state);
static regex_element_t *parse_atom(parser_state_t *state);

static char current_char(parser_state_t *state) {
    if (state->pos >= state->length) return '\0';
    return state->pattern[state->pos];
}

static void advance(parser_state_t *state) {
    if (state->pos < state->length) {
        state->pos++;
    }
}

static bool at_end(parser_state_t *state) {
    return state->pos >= state->length;
}

static bool consume(parser_state_t *state, char expected) {
    if (current_char(state) == expected) {
        advance(state);
        return true;
    }
    return false;
}

// Expression := Term ('|' Term)*
static regex_element_t *parse_expression(parser_state_t *state) {
    regex_element_t *left = parse_term(state);
    if (!left) return NULL;
    
    while (current_char(state) == '|') {
        advance(state); // consume '|'
        regex_element_t *right = parse_term(state);
        if (!right) {
            left->destroy(left);
            return NULL;
        }
        left = alternation_create(left, right);
        if (!left) {
            right->destroy(right);
            return NULL;
        }
    }
    
    return left;
}

// Term := Factor*
static regex_element_t *parse_term(parser_state_t *state) {
    regex_element_t *result = NULL;
    
    while (!at_end(state) && current_char(state) != '|' && current_char(state) != ')') {
        regex_element_t *factor = parse_factor(state);
        if (!factor) {
            if (result) result->destroy(result);
            return NULL;
        }
        
        if (result) {
            result = concat_create(result, factor);
            if (!result) {
                factor->destroy(factor);
                return NULL;
            }
        } else {
            result = factor;
        }
    }
    
    return result;
}

// Factor := Atom ('*' | '+' | '?')?
static regex_element_t *parse_factor(parser_state_t *state) {
    regex_element_t *atom = parse_atom(state);
    if (!atom) return NULL;
    
    char c = current_char(state);
    switch (c) {
        case '*':
            advance(state);
            return kleene_star_create(atom);
        case '+':
            advance(state);
            return plus_create(atom);
        case '?':
            advance(state);
            return question_create(atom);
        default:
            return atom;
    }
}

// Atom := CHAR | '(' Expression ')' | '.' | '\' EscapeChar
static regex_element_t *parse_atom(parser_state_t *state) {
    char c = current_char(state);
    
    switch (c) {
        case '\0':
            *(state->error) = FLOWREGEX_ERROR_PARSE;
            return NULL;
            
        case '(':
            advance(state); // consume '('
            {
                regex_element_t *expr = parse_expression(state);
                if (!expr) return NULL;
                
                if (!consume(state, ')')) {
                    expr->destroy(expr);
                    *(state->error) = FLOWREGEX_ERROR_PARSE;
                    return NULL;
                }
                return expr;
            }
            
        case '.':
            advance(state);
            return any_char_create();
            
        case '\\':
            advance(state); // consume '\'
            {
                char escape_char = current_char(state);
                if (escape_char == '\0') {
                    *(state->error) = FLOWREGEX_ERROR_PARSE;
                    return NULL;
                }
                advance(state);
                
                // Handle character classes
                switch (escape_char) {
                    case 'd':
                        return char_class_create("d");
                    case 'D':
                        return char_class_create("D");
                    case 's':
                        return char_class_create("s");
                    case 'S':
                        return char_class_create("S");
                    case 'w':
                        return char_class_create("w");
                    case 'W':
                        return char_class_create("W");
                    case 'n':
                        return literal_create('\n');
                    case 't':
                        return literal_create('\t');
                    case 'r':
                        return literal_create('\r');
                    case '\\':
                        return literal_create('\\');
                    case '.':
                        return literal_create('.');
                    default:
                        return literal_create(escape_char);
                }
            }
            
        case '|':
        case ')':
        case '*':
        case '+':
        case '?':
            *(state->error) = FLOWREGEX_ERROR_PARSE;
            return NULL;
            
        default:
            advance(state);
            return literal_create(c);
    }
}

regex_element_t *parse_regex(const char *pattern, flowregex_error_t *error) {
    if (!pattern || !error) {
        if (error) *error = FLOWREGEX_ERROR_INVALID_PATTERN;
        return NULL;
    }
    
    *error = FLOWREGEX_OK;
    
    parser_state_t state = {
        .pattern = pattern,
        .pos = 0,
        .length = strlen(pattern),
        .error = error
    };
    
    if (state.length == 0) {
        *error = FLOWREGEX_ERROR_INVALID_PATTERN;
        return NULL;
    }
    
    regex_element_t *result = parse_expression(&state);
    
    if (result && !at_end(&state)) {
        result->destroy(result);
        *error = FLOWREGEX_ERROR_PARSE;
        return NULL;
    }
    
    return result;
}
